### ISSUE TITLE
[BUG] Users are duplicated and cannot log in

### ISSUE THREAD

**Issue by MKlimuszka** [rancher/rancher#42082]
Rancher Server Setup

    Rancher version: 2.7.5
    Installation option (Helm Chart): K8S 1.24.x RKE1
    Proxy/Cert Details:

Information about the Cluster

    Kubernetes version: 1.24.x
    Cluster Type (Local/Downstream): Local

User Information

    What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
        All users are affected!

Describe the bug
After upgrading to Rancher 2.7.5 (from 2.7.4) LDAP based login using Active Directory integration fails for all users.

To Reproduce
Upgrade to 2.7.5 and then login with a fresh browser.

Result
Login failure. In Rachner log (GUID data removed):
LDAP Result Code 201 Filter Compile Error: ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E N

Expected Result
Login success

Additional context
The following change has introduced to problem:
https://github.com/rancher/rancher/commit/cdd90ea6c687d8c83cfd0289e51be32da5c21d14

I think the problem source is that guidString := html.EscapeString(fmt.Sprintf("%x", entry.GetRawAttributeValue(common.AttributeObjectGUID))) is used to build an URL from the objectGUID value. This might result in a string that is not fully hex encoded. All our objectGUIDs contain the "N" byte with hex value '4E'. The HTML escaping might not escape this byte and then when it is used later in LDAP the filter string will contain "...\N..." which is not a proper escaping and leads to the above noted error message. This is still an assumption, because I have not understood the relevant code fully yet, but I am pretty sure, that it points at least in the right direction.

**Comment by MKlimuszka** [comment]
This is an offshoot of https://github.com/rancher/rancher/issues/41985. There is a user duplication issue that cannot be reproduced yet. 

If the migration process fails to do anything for a user in Rancher, the next time that user logs-in, it will likely succeed, but since the new principalId for the user (which will be GUID-based in 2.7.5) will not match any existing principalIds (DN-based), Rancher will create a new user. This is the result we're looking to avoid when a user is moved from one OU to another (the original problem being solved by migrating to GUID-based principals), but given the migration failure, we can imagine a scenario where that happens.

For the case where customers do have duplicate user accounts there is a workaround:

Every user who logs in since 2.7.5 gets a 2nd user account created.
From the newer account copy the GUID under "principalIds" out and paste this into the "original account" (is a list, there is then the old DistinguishedName and additionally the new id in the format "activedirectory_user://f23114bd14ee2f4b9824d9e509XXXXXX" in it.
After that delete the newly created account - when the user logs out and logs in again he gets his old profile with all settings and tokens he ever created.

Hopefully this can help to "repair" the above situation.

So far, we have not managed to reproduce this outside of manually duplicating the user to see what the code was doing once it was in this state. Once we created that state, I could see why logins were failing (because there were 2 rancher users mapped to the same auth provider user). I would like to see a system or logs from a system that wound-up in that state because it's not yet obvious to me how it happens.

**Comment by sevko21** [comment]
Hi, we are in the same situation and can provide logs. As no one was able to login we did a rollback to 2.7.4 but this helped only partly.

**Comment by sevko21** [comment]
> This is an offshoot of #41985. There is a user duplication issue that cannot be reproduced yet.
> 
> If the migration process fails to do anything for a user in Rancher, the next time that user logs-in, it will likely succeed, but since the new principalId for the user (which will be GUID-based in 2.7.5) will not match any existing principalIds (DN-based), Rancher will create a new user. This is the result we're looking to avoid when a user is moved from one OU to another (the original problem being solved by migrating to GUID-based principals), but given the migration failure, we can imagine a scenario where that happens.
> 
> For the case where customers do have duplicate user accounts there is a workaround:
> 
> Every user who logs in since 2.7.5 gets a 2nd user account created. From the newer account copy the GUID under "principalIds" out and paste this into the "original account" (is a list, there is then the old DistinguishedName and additionally the new id in the format "activedirectory_user://f23114bd14ee2f4b9824d9e509XXXXXX" in it. After that delete the newly created account - when the user logs out and logs in again he gets his old profile with all settings and tokens he ever created.
> 
> Hopefully this can help to "repair" the above situation.
> 
> So far, we have not managed to reproduce this outside of manually duplicating the user to see what the code was doing once it was in this state. Once we created that state, I could see why logins were failing (because there were 2 rancher users mapped to the same auth provider user). I would like to see a system or logs from a system that wound-up in that state because it's not yet obvious to me how it happens.

does this also work on the gui? just tested with edit user and I dont see a principalid

**Comment by sowmyav27** [comment]
After offline conversation with @crobby This issue is fixed with PRs linked here - https://github.com/rancher/rancher/issues/41985

**Comment by crobby** [comment]
There is a script (currently PR) https://github.com/rancherlabs/support-tools/pull/232 that should address any users that are currently stuck with duplicate users.  
The current 2.7.5 will still break again though due to bugs there.  Those bugs have been fixed in the latest head as of July 19th.

**Comment by prachidamle** [comment]
The original migration in v2.7.5 lead to duplicate users and users who cannot login. That has been all reverted in v2.7.6, so moving this to-test.

Revert tickets:
https://github.com/rancher/rancher/issues/42174
https://github.com/rancher/rancher/issues/42161

**Comment by Priyashetty17** [comment]
Re-opening this since un-migration logic does not seem to have been merged to 2.7 yet.

I ran a upgrade usecase from 2.7.5 to 2.7.7-rc4. After upgrade, I still see the duplicate user issue. Also, I do not see any migration related configmap or log messages in the rancher log. 

**Steps to Recreate:**
```
1. Start with Rancher 2.7.5.
2. Enable AD 
3. Create a downstream cluster and add few AD users to the cluster as CO. 
4. Login as the AD users. 
5. Upgrade from 2.7.5 to 2.7.7.-rc4
6. Login as the AD users. 
```

**Issues:**
```
- Duplicate AD users 
- AD users ^^ are able to login but cannot view any cluster after login.
- AD users ^^ have both GUID and DN instead of only DN
- Config map ad-guid-migration does not exist
	kubectl get cm ad-guid-migration -n cattle-system -o yaml
	Error from server (NotFound): configmaps "ad-guid-migration" not found
- CRTBs have GUID for userPrincipalName instead of DN.
- Rancher log does not have unmigration related messages. 
```
